substrate/client/db/src/cache.rs

```diff
@@ -0,0 +1,412 @@
+// Copyright 2017 Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! DB-backed cache of blockchain data.
+
+use std::sync::Arc;
+use parking_lot::RwLock;
+
+use kvdb::{KeyValueDB, DBTransaction};
+
+use client::blockchain::Cache as BlockchainCache;
+use client::error::Result as ClientResult;
+use codec::{Codec, Encode, Decode, Input, Output};
+use primitives::AuthorityId;
+use runtime_primitives::generic::BlockId;
+use runtime_primitives::traits::{Block as BlockT, Header as HeaderT, As};
+use utils::{COLUMN_META, BlockKey, db_err, meta_keys, read_id, db_key_to_number, number_to_db_key};
+
+/// Database-backed cache of blockchain data.
+pub struct DbCache<Block: BlockT> {
+	db: Arc<KeyValueDB>,
+	block_index_column: Option<u32>,
+	authorities_at: DbCacheList<Block, Vec<AuthorityId>>,
+}
+
+impl<Block> DbCache<Block>
+	where
+		Block: BlockT,
+		<<Block as BlockT>::Header as HeaderT>::Number: As<u64>,
+{
+	/// Create new cache.
+	pub fn new(db: Arc<KeyValueDB>, block_index_column: Option<u32>, authorities_column: Option<u32>) -> ClientResult<Self> {
+		Ok(DbCache {
+			db: db.clone(),
+			block_index_column,
+			authorities_at: DbCacheList::new(db, meta_keys::BEST_AUTHORITIES, authorities_column)?,
+		})
+	}
+
+	/// Get authorities_cache.
+	pub fn authorities_at_cache(&self) -> &DbCacheList<Block, Vec<AuthorityId>> {
+		&self.authorities_at
+	}
+}
+
+impl<Block> BlockchainCache<Block> for DbCache<Block>
+	where
+		Block: BlockT,
+		<<Block as BlockT>::Header as HeaderT>::Number: As<u64>,
+{
+	fn authorities_at(&self, at: BlockId<Block>) -> Option<Vec<AuthorityId>> {
+		let authorities_at = read_id(&*self.db, self.block_index_column, at).and_then(|at| match at {
+			Some(at) => self.authorities_at.value_at_key(at),
+			None => Ok(None),
```

Comment:
My guess - when we're calling `authorities_at(BlockId::Hash(unknown_hash))`. Or am I wrong? :) 
Anyway - it is  the cache, I'd prefer to write cache code without any extra `expects`, since this is an optimization anyway.